// @ocavue/utils/src/dom-node-type.ts
// prettier-ignore
export const ELEMENT_NODE = 1 satisfies typeof Node.ELEMENT_NODE;
// prettier-ignore
export const ATTRIBUTE_NODE = 2 satisfies typeof Node.ATTRIBUTE_NODE;
// prettier-ignore
export const TEXT_NODE = 3 satisfies typeof Node.TEXT_NODE;
// prettier-ignore
export const CDATA_SECTION_NODE = 4 satisfies typeof Node.CDATA_SECTION_NODE;
// prettier-ignore
export const ENTITY_REFERENCE_NODE = 5 satisfies typeof Node.ENTITY_REFERENCE_NODE;
// prettier-ignore
export const ENTITY_NODE = 6 satisfies typeof Node.ENTITY_NODE;
// prettier-ignore
export const PROCESSING_INSTRUCTION_NODE = 7 satisfies typeof Node.PROCESSING_INSTRUCTION_NODE;
// prettier-ignore
export const COMMENT_NODE = 8 satisfies typeof Node.COMMENT_NODE;
// prettier-ignore
export const DOCUMENT_NODE = 9 satisfies typeof Node.DOCUMENT_NODE;
// prettier-ignore
export const DOCUMENT_TYPE_NODE = 10 satisfies typeof Node.DOCUMENT_TYPE_NODE;
// prettier-ignore
export const DOCUMENT_FRAGMENT_NODE = 11 satisfies typeof Node.DOCUMENT_FRAGMENT_NODE;
// prettier-ignore
export const NOTATION_NODE = 12 satisfies typeof Node.NOTATION_NODE;
